src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,326 @@
+
+namespace {
+kj::Maybe<AsyncContextFrame&> tryGetContextFrame(
+    v8::Isolate* isolate,
+    v8::Local<v8::Value> handle) {
+  // Gets the held AsyncContextFrame from the opaque wrappable but does not consume it.
+  KJ_IF_MAYBE(wrappable, Wrappable::tryUnwrapOpaque(isolate, handle)) {
+    AsyncContextFrame* frame = dynamic_cast<AsyncContextFrame*>(wrappable);
+    KJ_ASSERT(frame != nullptr);
+    return *frame;
+  }
+  return nullptr;
+}
+
+}  // namespace
+
+AsyncContextFrame::AsyncContextFrame(IsolateBase& isolate)
+    : isolate(isolate) {}
+
+AsyncContextFrame::AsyncContextFrame(
+    Lock& js,
+    kj::Maybe<AsyncContextFrame&> maybeParent,
+    kj::Maybe<StorageEntry> maybeStorageEntry)
+    : AsyncContextFrame(IsolateBase::from(js.v8Isolate)) {
+
+  // Propagate the storage context of the parent frame to this newly created frame.
+  const auto propagate = [&](AsyncContextFrame& parent) {
+    parent.storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
+    for (auto& entry : parent.storage) {
+      storage.insert(entry.clone(js));
+    }
+
+    KJ_IF_MAYBE(entry, maybeStorageEntry) {
+      storage.upsert(kj::mv(*entry), [](StorageEntry& existing, StorageEntry&& row) mutable {
+        existing.value = kj::mv(row.value);
+      });
+    }
+  };
+
+  KJ_IF_MAYBE(parent, maybeParent) {
+    propagate(*parent);
+  } else {
+    propagate(current(js));
+  }
+}
+
+kj::Maybe<AsyncContextFrame&> AsyncContextFrame::tryGetContext(
+    Lock& js,
+    v8::Local<v8::Promise> promise) {
+  auto handle = js.getPrivateSymbolFor(Lock::PrivateSymbols::ASYNC_RESOURCE);
+  // We do not use the normal unwrapOpaque here since that would consume the wrapped
+  // value, and we need to be able to unwrap multiple times.
+  return tryGetContextFrame(js.v8Isolate,
+      check(promise->GetPrivate(js.v8Isolate->GetCurrentContext(), handle)));
+}
+
+kj::Maybe<AsyncContextFrame&> AsyncContextFrame::tryGetContext(
+    Lock& js,
+    V8Ref<v8::Promise>& promise) {
+  return tryGetContext(js, promise.getHandle(js));
+}
+
+AsyncContextFrame& AsyncContextFrame::current(Lock& js) {
+  auto& isolateBase = IsolateBase::from(js.v8Isolate);
+  if (isolateBase.asyncFrameStack.size() == 0) {
+    return *isolateBase.getRootAsyncContext();
+  }
+  return *isolateBase.asyncFrameStack.back();
+}
+
+Ref<AsyncContextFrame> AsyncContextFrame::create(
+    Lock& js,
+    kj::Maybe<AsyncContextFrame&> maybeParent,
+    kj::Maybe<StorageEntry> maybeStorageEntry) {
+  return alloc<AsyncContextFrame>(js, maybeParent, kj::mv(maybeStorageEntry));
+}
+
+v8::Local<v8::Function> AsyncContextFrame::wrap(
+    Lock& js,
+    v8::Local<v8::Function> fn,
+    kj::Maybe<AsyncContextFrame&> maybeFrame,
+    kj::Maybe<v8::Local<v8::Value>> thisArg) {
+  auto isolate = js.v8Isolate;
+  auto context = isolate->GetCurrentContext();
+  return js.wrapReturningFunction(context, JSG_VISITABLE_LAMBDA(
+      (
+        frame = kj::addRef(AsyncContextFrame::current(js)),
```

Comment:
Since this inherits `Wrappable` I think you need to be capturing a `jsg::Ref` rather than a `kj::addRef`. Plus then you could add it to the GC visitation list.